# Neural Network Assignment- Forest Fires

## Importing Libraries

In [1]:
!pip install keras

In [1]:
# Importing the necessary packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Loading Dataset

In [2]:
Fire_Forest = pd.read_csv("forestfires.csv")
Fire_Forest.head()

month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  ...  monthfeb  \
0   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0  ...         0   
1   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0  ...         0   
2   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0  ...         0   
3   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2  ...         0   
4   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0  ...         0   

   monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0         0         0         0         1         0         0         0   
1         0         0         0         0         0         0         1   
2         0         0         0         0         0         0         1   
3         0         0         0         1         0         0         0   
4         0         0         0         1         0         0         0   

   monthsep  size_category  
0         0          small  
1         0          small  
2         0          small  
3         0          small  
4         0          small  

[5 rows x 31 columns]

## data Understanding

In [3]:
Fire_Forest.shape

(517, 31)

In [4]:
Fire_Forest.dtypes

month             object
day               object
FFMC             float64
DMC              float64
DC               float64
ISI              float64
temp             float64
RH                 int64
wind             float64
rain             float64
area             float64
dayfri             int64
daymon             int64
daysat             int64
daysun             int64
daythu             int64
daytue             int64
daywed             int64
monthapr           int64
monthaug           int64
monthdec           int64
monthfeb           int64
monthjan           int64
monthjul           int64
monthjun           int64
monthmar           int64
monthmay           int64
monthnov           int64
monthoct           int64
monthsep           int64
size_category     object
dtype: object

In [5]:
Fire_Forest.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 31 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   month          517 non-null    object 
 1   day            517 non-null    object 
 2   FFMC           517 non-null    float64
 3   DMC            517 non-null    float64
 4   DC             517 non-null    float64
 5   ISI            517 non-null    float64
 6   temp           517 non-null    float64
 7   RH             517 non-null    int64  
 8   wind           517 non-null    float64
 9   rain           517 non-null    float64
 10  area           517 non-null    float64
 11  dayfri         517 non-null    int64  
 12  daymon         517 non-null    int64  
 13  daysat         517 non-null    int64  
 14  daysun         517 non-null    int64  
 15  daythu         517 non-null    int64  
 16  daytue         517 non-null    int64  
 17  daywed         517 non-null    int64  
 18  monthapr  

In [6]:
Fire_Forest.isnull().sum()

month            0
day              0
FFMC             0
DMC              0
DC               0
ISI              0
temp             0
RH               0
wind             0
rain             0
area             0
dayfri           0
daymon           0
daysat           0
daysun           0
daythu           0
daytue           0
daywed           0
monthapr         0
monthaug         0
monthdec         0
monthfeb         0
monthjan         0
monthjul         0
monthjun         0
monthmar         0
monthmay         0
monthnov         0
monthoct         0
monthsep         0
size_category    0
dtype: int64

## There is not any null value in a dataset

In [7]:
Fire_Forest.describe()

FFMC         DMC          DC         ISI        temp          RH  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean    90.644681  110.872340  547.940039    9.021663   18.889168   44.288201   
std      5.520111   64.046482  248.066192    4.559477    5.806625   16.317469   
min     18.700000    1.100000    7.900000    0.000000    2.200000   15.000000   
25%     90.200000   68.600000  437.700000    6.500000   15.500000   33.000000   
50%     91.600000  108.300000  664.200000    8.400000   19.300000   42.000000   
75%     92.900000  142.400000  713.900000   10.800000   22.800000   53.000000   
max     96.200000  291.300000  860.600000   56.100000   33.300000  100.000000   

             wind        rain         area      dayfri  ...    monthdec  \
count  517.000000  517.000000   517.000000  517.000000  ...  517.000000   
mean     4.017602    0.021663    12.847292    0.164410  ...    0.017408   
std      1.791653    0.295959    63.655818    0.371006  ...    0.130913   
min      0.400000    0.000000     0.000000    0.000000  ...    0.000000   
25%      2.700000    0.000000     0.000000    0.000000  ...    0.000000   
50%      4.000000    0.000000     0.520000    0.000000  ...    0.000000   
75%      4.900000    0.000000     6.570000    0.000000  ...    0.000000   
max      9.400000    6.400000  1090.840000    1.000000  ...    1.000000   

         monthfeb    monthjan    monthjul    monthjun    monthmar    monthmay  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean     0.038685    0.003868    0.061896    0.032882    0.104449    0.003868   
std      0.193029    0.062137    0.241199    0.178500    0.306138    0.062137   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

         monthnov    monthoct    monthsep  
count  517.000000  517.000000  517.000000  
mean     0.001934    0.029014    0.332689  
std      0.043980    0.168007    0.471632  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000  
75%      0.000000    0.000000    1.000000  
max      1.000000    1.000000    1.000000  

[8 rows x 28 columns]

## Using Label Encoding

In [8]:
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

In [9]:
LabelEncoder = preprocessing.LabelEncoder()
Fire_Forest['month'] = LabelEncoder.fit_transform(Fire_Forest['month'])
Fire_Forest['day'] = LabelEncoder.fit_transform(Fire_Forest['day'])
Fire_Forest['size_category'] = LabelEncoder.fit_transform(Fire_Forest['size_category'])

In [10]:
Fire_Forest.head()

month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  ...  monthfeb  \
0      7    0  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0  ...         0   
1     10    5  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0  ...         0   
2     10    2  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0  ...         0   
3      7    0  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2  ...         0   
4      7    3  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0  ...         0   

   monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0         0         0         0         1         0         0         0   
1         0         0         0         0         0         0         1   
2         0         0         0         0         0         0         1   
3         0         0         0         1         0         0         0   
4         0         0         0         1         0         0         0   

   monthsep  size_category  
0         0              1  
1         0              1  
2         0              1  
3         0              1  
4         0              1  

[5 rows x 31 columns]

## Splitting a data into features and target

In [11]:
X = Fire_Forest.iloc[:,0:11]
y = Fire_Forest.iloc[:,-1]

In [12]:
X

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain   area
0        7    0  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.00
1       10    5  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.00
2       10    2  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.00
3        7    0  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.00
4        7    3  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.00
..     ...  ...   ...    ...    ...   ...   ...  ..   ...   ...    ...
512      1    3  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0   6.44
513      1    3  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  54.29
514      1    3  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  11.16
515      1    2  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0   0.00
516      9    5  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0   0.00

[517 rows x 11 columns]

In [13]:
y

0      1
1      1
2      1
3      1
4      1
      ..
512    0
513    0
514    0
515    1
516    1
Name: size_category, Length: 517, dtype: int32

In [14]:
y = pd.DataFrame(y)

In [15]:
y

size_category
0                1
1                1
2                1
3                1
4                1
..             ...
512              0
513              0
514              0
515              1
516              1

[517 rows x 1 columns]

## Scaling a data using Standard Scaler

In [16]:
from sklearn.preprocessing import StandardScaler
std_scalar =  StandardScaler()

In [17]:
scaled_x = std_scalar.fit_transform(X)

In [18]:
scaled_x

array([[ 0.28422225, -1.42312073, -0.80595947, ...,  1.49861442,
        -0.07326831, -0.20201979],
       [ 0.97087134,  1.17671466, -0.00810203, ..., -1.74175564,
        -0.07326831, -0.20201979],
       [ 0.97087134, -0.38318657, -0.00810203, ..., -1.51828184,
        -0.07326831, -0.20201979],
       ...,
       [-1.08907592,  0.13678051, -1.64008316, ...,  1.49861442,
        -0.07326831, -0.02653216],
       [-1.08907592, -0.38318657,  0.68095666, ..., -0.00983371,
        -0.07326831, -0.20201979],
       [ 0.74198831,  1.17671466, -2.02087875, ...,  0.26950853,
        -0.07326831, -0.20201979]])

## Model Evaluation

## Tensorflow is an important library used in Neural Network

In [20]:
!pip install tensorflow

## Sequential model is used for model building

In [21]:
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense, Activation, Layer, Lambda

In [22]:
from sklearn.model_selection import train_test_split

In [23]:
X_train, X_test, y_train, y_test = train_test_split(scaled_x,y, test_size = 0.3, random_state=42)

In [24]:
X_train.shape, X_test.shape

((361, 11), (156, 11))

In [25]:
y_train.shape, y_test.shape

((361, 1), (156, 1))

In [26]:
import tensorflow as tf
model = Sequential()
model.add(Dense(units=20, activation='relu'))
model.add(Dense(units=20, activation='tanh'))
model.add(Dense(units=20, activation='softmax'))

In [27]:
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

## Optimizers are used to avoid Vanishing/Exploding gradient in back propagation

In [28]:
model.fit(x=X_train, y=y_train, epochs=517)

Epoch 1/517
12/12 [==============================] - 1s 2ms/step - loss: 2.9321 - accuracy: 0.0776
Epoch 2/517
12/12 [==============================] - 0s 2ms/step - loss: 2.6955 - accuracy: 0.3102
Epoch 3/517
12/12 [==============================] - 0s 2ms/step - loss: 2.4528 - accuracy: 0.6011
Epoch 4/517
12/12 [==============================] - 0s 2ms/step - loss: 2.1916 - accuracy: 0.7064
Epoch 5/517
12/12 [==============================] - 0s 3ms/step - loss: 1.9044 - accuracy: 0.7285
Epoch 6/517
12/12 [==============================] - 0s 2ms/step - loss: 1.6118 - accuracy: 0.7313
Epoch 7/517
12/12 [==============================] - 0s 2ms/step - loss: 1.3432 - accuracy: 0.7285
Epoch 8/517
12/12 [==============================] - 0s 3ms/step - loss: 1.1263 - accuracy: 0.7285
Epoch 9/517
12/12 [==============================] - 0s 2ms/step - loss: 0.9660 - accuracy: 0.7285
Epoch 10/517
12/12 [==============================] - 0s 2ms/step - loss: 0.8553 - accuracy: 0.7285
Epoch 11/

12/12 [==============================] - 0s 2ms/step - loss: 0.1813 - accuracy: 0.9418
Epoch 84/517
12/12 [==============================] - 0s 3ms/step - loss: 0.1750 - accuracy: 0.9391
Epoch 85/517
12/12 [==============================] - 0s 3ms/step - loss: 0.1696 - accuracy: 0.9501
Epoch 86/517
12/12 [==============================] - 0s 3ms/step - loss: 0.1651 - accuracy: 0.9501
Epoch 87/517
12/12 [==============================] - 0s 2ms/step - loss: 0.1604 - accuracy: 0.9557
Epoch 88/517
12/12 [==============================] - 0s 3ms/step - loss: 0.1559 - accuracy: 0.9529
Epoch 89/517
12/12 [==============================] - 0s 3ms/step - loss: 0.1516 - accuracy: 0.9584
Epoch 90/517
12/12 [==============================] - 0s 2ms/step - loss: 0.1475 - accuracy: 0.9612
Epoch 91/517
12/12 [==============================] - 0s 2ms/step - loss: 0.1443 - accuracy: 0.9612
Epoch 92/517
12/12 [==============================] - 0s 2ms/step - loss: 0.1390 - accuracy: 0.9640
Epoch 93/517


12/12 [==============================] - 0s 3ms/step - loss: 0.0086 - accuracy: 1.0000
Epoch 246/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0087 - accuracy: 1.0000
Epoch 247/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0086 - accuracy: 1.0000
Epoch 248/517
12/12 [==============================] - 0s 3ms/step - loss: 0.0085 - accuracy: 1.0000
Epoch 249/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0083 - accuracy: 1.0000
Epoch 250/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0080 - accuracy: 1.0000
Epoch 251/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0079 - accuracy: 1.0000
Epoch 252/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0078 - accuracy: 1.0000
Epoch 253/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0079 - accuracy: 1.0000
Epoch 254/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0076 - accuracy: 1.0000
Epoc

12/12 [==============================] - 0s 2ms/step - loss: 0.0016 - accuracy: 1.0000
Epoch 408/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 409/517
12/12 [==============================] - 0s 3ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 410/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 411/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 412/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 413/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 414/517
12/12 [==============================] - 0s 3ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 415/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0015 - accuracy: 1.0000
Epoch 416/517
12/12 [==============================] - 0s 2ms/step - loss: 0.0014 - accuracy: 1.0000
Epoc

In [29]:
y_pred = model.predict(X_test)

In [30]:
y_pred

array([[1.5352225e-05, 9.9997735e-01, 9.8520002e-07, ..., 3.5769415e-07,
        2.2795085e-07, 6.1942666e-07],
       [1.9578569e-05, 9.9997604e-01, 4.4422137e-07, ..., 2.1942350e-07,
        1.2559103e-07, 4.5420100e-07],
       [7.5545072e-06, 9.9998879e-01, 5.0507941e-07, ..., 1.9163011e-07,
        1.1002271e-07, 2.9056790e-07],
       ...,
       [3.7821424e-05, 9.9994993e-01, 1.5226082e-06, ..., 6.5198293e-07,
        3.9833634e-07, 1.3247372e-06],
       [3.5770025e-02, 9.6375960e-01, 7.0601185e-05, ..., 2.3550647e-05,
        1.4640077e-05, 4.0713414e-05],
       [6.8567113e-07, 9.9999666e-01, 4.6464609e-07, ..., 1.6348815e-07,
        8.3285570e-08, 1.6854595e-07]], dtype=float32)

In [31]:
y_pred = pd.DataFrame(y_pred)
y_pred

0         1             2             3             4   \
0    1.535223e-05  0.999977  9.852000e-07  3.459301e-07  5.393927e-07   
1    1.957857e-05  0.999976  4.442214e-07  2.740232e-07  2.998111e-07   
2    7.554507e-06  0.999989  5.050794e-07  1.930158e-07  3.221651e-07   
3    4.712195e-04  0.999488  5.666597e-06  2.066879e-06  2.775686e-06   
4    5.609729e-05  0.999933  1.448059e-06  5.872221e-07  8.327639e-07   
..            ...       ...           ...           ...           ...   
151  9.857886e-01  0.013552  7.821407e-05  2.774370e-05  3.086638e-05   
152  1.389583e-05  0.999979  1.037414e-06  3.207451e-07  5.673976e-07   
153  3.782142e-05  0.999950  1.522608e-06  5.817751e-07  7.610184e-07   
154  3.577003e-02  0.963760  7.060119e-05  2.238214e-05  2.519654e-05   
155  6.856711e-07  0.999997  4.646461e-07  9.578689e-08  3.366603e-07   

               5             6             7             8             9   \
0    2.417645e-07  2.497886e-07  3.703209e-07  6.220486e-07  3.943854e-07   
1    1.048182e-07  1.502126e-07  2.245312e-07  3.292439e-07  1.880726e-07   
2    1.076406e-07  1.176425e-07  1.699724e-07  3.107380e-07  1.757709e-07   
3    1.096569e-06  1.830144e-06  2.020464e-06  2.757917e-06  2.599538e-06   
4    3.247269e-07  4.482581e-07  5.288808e-07  8.787918e-07  6.240975e-07   
..            ...           ...           ...           ...           ...   
151  2.738601e-05  4.098716e-05  3.208100e-05  3.903987e-05  5.328368e-05   
152  1.905655e-07  2.381330e-07  3.242351e-07  5.385632e-07  3.539938e-07   
153  4.039253e-07  4.593853e-07  6.734179e-07  8.675253e-07  7.220325e-07   
154  1.867722e-05  2.305790e-05  2.569735e-05  2.688937e-05  2.843876e-05   
155  6.932552e-08  6.661136e-08  1.190417e-07  2.922073e-07  1.067580e-07   

               10            11            12            13            14  \
0    2.488378e-07  1.860617e-07  3.530363e-07  1.869869e-07  2.457353e-07   
1    1.058759e-07  1.083022e-07  2.314347e-07  1.253020e-07  1.173760e-07   
2    9.285434e-08  8.323585e-08  1.816719e-07  9.316497e-08  1.089972e-07   
3    1.237438e-06  1.071167e-06  2.254326e-06  1.152423e-06  1.067461e-06   
4    3.456875e-07  2.571703e-07  6.901085e-07  2.776419e-07  3.056572e-07   
..            ...           ...           ...           ...           ...   
151  2.404656e-05  2.026409e-05  5.512933e-05  1.912203e-05  1.299881e-05   
152  1.955598e-07  1.568364e-07  3.237062e-07  1.729963e-07  2.058752e-07   
153  4.443674e-07  3.803458e-07  5.756166e-07  3.477424e-07  3.993820e-07   
154  1.532274e-05  1.519589e-05  2.744944e-05  1.285249e-05  1.075742e-05   
155  7.989689e-08  6.165222e-08  1.069122e-07  5.467827e-08  8.675556e-08   

               15            16            17            18            19  
0    6.148161e-07  5.111204e-07  3.576941e-07  2.279508e-07  6.194267e-07  
1    3.715961e-07  4.296378e-07  2.194235e-07  1.255910e-07  4.542010e-07  
2    3.162956e-07  2.920946e-07  1.916301e-07  1.100227e-07  2.905679e-07  
3    3.353136e-06  3.249959e-06  1.747789e-06  1.230530e-06  3.311042e-06  
4    1.041164e-06  9.311788e-07  4.763667e-07  3.030201e-07  8.789510e-07  
..            ...           ...           ...           ...           ...  
151  4.146687e-05  4.845643e-05  2.914085e-05  1.853730e-05  6.105032e-05  
152  5.848853e-07  4.873226e-07  3.107008e-07  2.024145e-07  4.668380e-07  
153  9.250978e-07  8.030677e-07  6.519829e-07  3.983363e-07  1.324737e-06  
154  3.664506e-05  3.224609e-05  2.355065e-05  1.464008e-05  4.071341e-05  
155  1.791153e-07  1.976655e-07  1.634882e-07  8.328557e-08  1.685460e-07  

[156 rows x 20 columns]

In [32]:
FireForest_train = model.evaluate(X_train, y_train)
FireForest_train 

12/12 [==============================] - 0s 3ms/step - loss: 5.9304e-04 - accuracy: 1.0000


[0.0005930443876422942, 1.0]

In [33]:
FireForest_test = model.evaluate(X_test, y_test)
FireForest_test

5/5 [==============================] - 0s 2ms/step - loss: 0.1448 - accuracy: 0.9615


[0.14483115077018738, 0.9615384340286255]

# Conclusion:- Test data is giving an accuracy of 96.15%